kcrypt/encryptor.go

```diff
@@ -0,0 +1,735 @@
+}
+
+func (e *LocalTPMNVEncryptor) Name() string {
+	return "Local TPM NV passphrase"
+}
+
+func (e *LocalTPMNVEncryptor) Validate() error {
+	// Validate TPM 2.0 device exists
+	return validateTPMDevice(e.logger)
+}
+
+// GetEncryptor returns the appropriate encryptor based on system configuration.
+// It automatically:
+// 1. Scans for kcrypt configuration (challenger server, mdns, etc.)
+// 2. Detects UKI mode
+// 3. Extracts PCR bindings from config
+// 4. Returns the appropriate encryptor implementation
+//
+// Decision logic:
+// 1. If challenger_server configured OR mdns enabled -> Remote KMS
+// 2. Else if UKI mode -> TPM + PCR policy (requires systemd >= 252 and TPM 2.0)
+// 3. Else (non-UKI, no remote) -> Local TPM NV passphrase.
+func GetEncryptor(logger types.KairosLogger) (PartitionEncryptor, error) {
+	kcryptConfig := ScanKcryptConfig(logger)
+
+	isUKI := detectUKIMode(logger)
+
+	var bindPCRs, bindPublicPCRs []string
+	if isUKI {
+		collectorConfig := scanCollectorConfig(logger)
+		if collectorConfig != nil {
+			bindPCRs, bindPublicPCRs = extractPCRBindingsFromCollector(*collectorConfig, logger)
+		}
+	}
+
+	useRemoteKMS := kcryptConfig != nil && (kcryptConfig.ChallengerServer != "" || kcryptConfig.MDNS)
+
+	var encryptor PartitionEncryptor
+
+	if useRemoteKMS {
+		logger.Logger.Info().
+			Str("challenger_server", kcryptConfig.ChallengerServer).
+			Bool("mdns", kcryptConfig.MDNS).
+			Msg("Using remote KMS for encryption")
+		encryptor = &RemoteKMSEncryptor{
+			logger:       logger,
+			kcryptConfig: kcryptConfig,
+		}
+	} else if isUKI {
+		logger.Logger.Info().Msg("Using TPM with PCR policy for encryption (UKI mode)")
+		encryptor = &TPMWithPCREncryptor{
+			logger:         logger,
+			bindPublicPCRs: bindPublicPCRs,
+			bindPCRs:       bindPCRs,
+		}
+	} else {
+		logger.Logger.Info().Msg("Using local TPM NV passphrase for encryption")
+		encryptor = &LocalTPMNVEncryptor{
+			logger:       logger,
+			kcryptConfig: kcryptConfig,
+		}
+	}
+
+	if err := encryptor.Validate(); err != nil {
+		return nil, fmt.Errorf("encryptor validation failed: %w", err)
+	}
+
+	return encryptor, nil
+}
+
+// scanCollectorConfig scans for configuration and returns the collector config.
+func scanCollectorConfig(logger types.KairosLogger) *collector.Config {
+	o := &collector.Options{NoLogs: true, MergeBootCMDLine: true}
+	if err := o.Apply(collector.Directories(DefaultConfigDirs...)); err != nil {
+		logger.Debugf("scanCollectorConfig: error applying collector options: %v", err)
+		return nil
+	}
+
+	collectorConfig, err := collector.Scan(o, func(d []byte) ([]byte, error) {
+		return d, nil
+	})
+	if err != nil {
+		logger.Debugf("scanCollectorConfig: error scanning for config: %v", err)
+		return nil
+	}
+
+	return collectorConfig
+}
+
+// detectUKIMode detects if the system is running in UKI mode
+// This checks for the presence of UKI-specific indicators.
+func detectUKIMode(logger types.KairosLogger) bool {
```

Comment:
sdk provides this under `state.DetectUKIboot` which just checks if cdmline contains the `rd.immucore.uki`

Reply:
I think I saw that but I don't remember why I ended up implementing this one. Given DetectUKIboot needs the cmdline as an argument, I will keep the wrapper method here but change it to use DetectUKIboot after reading the cmdline, like we do in the other calls of DetectUKIboot.

Reply:
:+1: I think its done that way to allow for testing to change the paths, so you can pass a fake cmdline to it to trigger the uki boot detection